In [7]:
import pandas as pd
from sklearn.model_selection import train_test_split

# Import the data and take a quick peak at it
df = pd.read_csv("../data/CTA_Ridership_Cleaned.csv")
df.head()

Unnamed: 0 service_date day_type     bus  rail_boardings  total_rides  \
0        1195   2001-01-01        U  297192          126455       423647   
1        7906   2001-01-02        W  780827          501952      1282779   
2        6604   2001-01-03        W  824923          536432      1361355   
3        6528   2001-01-04        W  870021          550011      1420032   
4        5067   2001-01-05        W  890426          557917      1448343   

   year  month    weekday  
0  2001      1     Monday  
1  2001      1    Tuesday  
2  2001      1  Wednesday  
3  2001      1   Thursday  
4  2001      1     Friday

In [16]:
df["service_date"] = pd.to_datetime(df["service_date"])
df['day'] = df["service_date"].dt.day
df["weekday"] = df["service_date"].dt.weekday
df["dayofyear"] = df["service_date"].dt.dayofyear
df["prev_day_rides"] = df["total_rides"].shift(1)
df = df.dropna()

In [23]:
# Define our X and our y so we can begin model training
X = df[["dayofyear", "year", "month", "day", "weekday", "prev_day_rides"]]
y = df["total_rides"]

# Create our test train and split
X_train, X_test, y_train, y_test = train_test_split(X, y, shuffle = False, test_size = 42)

# Now we can import our model, from what we say in EDA our data is not linear meaning
# we should use Random Forest and import MAE to eval the model

from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_absolute_error, r2_score

model = RandomForestRegressor(n_estimators = 100, random_state = 42)
model.fit(X_train, y_train)

y_pred = model.predict(X_test)
mae = mean_absolute_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"MAE: {mae:,.0f} Rides")
print(f"R2 Score: {r2:.2f}")

MAE: 81,419 Rides
R2 Score: 0.62


### As we can see we obtained an R2 score of 0.62 which is decent, but we can do better. Lets do some feature engineering to try to get a better score